enhancements/installer/vsphere-ipi-multiple-datacenter-clusters.md

```diff
@@ -0,0 +1,474 @@
+datacenters and clusters.
+
+### installer
+
+#### Platform spec
+
+The platform spec needs to be modified to support our initial goals of multiple
+datacenters (regions) and clusters (zones) and vcenters.
+
+We are adding one additional parameter to the Platform struct:
+
+- `VCenters`
+
+`VCenters` will contain the connection details for each individual vcenter
+and the `Region` associated with that `VCenter`. The
+existing platform spec vcenter parameters will be deprecated but _not_ removed
+or remove support for using those parameters.
+
+Regions will contain the datacenter and `Zones` that its associated with.
+Each region's zones will be associated with the
+additional configuration data required to provision virtual machines within
+vSphere. Including `Cluster`, `Network` (port group), `ResourcePool` and the
+`Datastore`.
+
+```golang
+type Platform struct {
+...
+  VCenters []VCenter `json:"vcenters,omitempty"`
+}
+
+type VCenter struct {
+  Server string `json:"server,omitempty"`
+  Port uint `json:"port,omitempty"`
+  User string `json:"user,omitempty"`
+  Password string `json:"password,omitempty"`
+  Regions []Region `json:"regions,omitempty"`
+}
+
+type Region struct {
+  Name string `json:"name,omitempty"`
+  Datacenter string `json:"datacenter,omitempty"`
+  Zones []Zone `json:"zones,omitempty"`
+}
+
+type Zone struct {
+  Name string `json:"name,omitempty"`
+  Cluster string `json:"cluster,omitempty"`
+  Network string `json:"network,omitempty"`
+  ResourcePool string `json:"resourcepool,omitempty"`
+  Datastore string `json:"datastore,omitempty"`
+}
+```
+
+Add [platform validation](https://github.com/openshift/installer/blob/master/pkg/types/vsphere/validation/platform.go)
+for the new struct fields that are required.
+
+#### MachinePool
+
+The `MachinePool` [struct](https://github.com/openshift/installer/blob/master/pkg/types/vsphere/machinepool.go#L5-L26)
+needs a single change to include the zones.
+The `Zones` like the other cloud providers will determine the location
+where the virtual machine will reside within a vSphere environment.
+
+```golang
+type MachinePool struct {
+...
+  Zones []string `json:"zones,omitempty"`
+}
+```
+
+Add [machinepool validation](https://github.com/openshift/installer/blob/master/pkg/types/vsphere/validation/machinepool.go)
+the zone name has a corresponding entry in `Regions.Zones`.
+
+#### Cloud Config
+
+The out-of-tree CCM is required for this change. The out-of-tree CCM also updates
+the cloud-config ini configuration. Below is an example of what the
+cloud-config will change to. A `VirtualCenter` section will be added per
+vcenter in `vcenters`. `datacenters` is a comma-delimited string that will contain
+all the datacenters per region.
+
+```ini
+[Global]
+  insecure-flag = "1"
+  secret-name = "vsphere-creds"
+  secret-namespace = "kube-system"
+
+[VirtualCenter "10.0.0.1"]
+  datacenters = "SDDC-Datacenter"
+
+[Labels]
+  region = openshift-region
```

Comment:
Should we store this information also in infrastructure object?
I'm bit concerned about possible changes of this config from user side, and potential data source for possible cluster-api provider landing.

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/docs/proposal/20201103-failure-domain.md#api-design - `FailureDomain` struct there

Reply:
We kinda have this problem today right? A customer could go in and mess up their config.

Reply:
Having this information within the infrastructure object would make it easier for components to discover rather than having to have the installer spread it around the cluster. Eg, we will probably want CAPI to have access to this information in the future too

Reply:
@lobziik @JoelSpeed can you be more specific what you are looking for?

https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go
The cloud-config is already apart of the InfrastructureSpec 

CAPI isn't in scope for this effort.

Reply:
So we have two options in the future:

1) We store the zonal information only in the cloud conf - any project that wishes to consume this information will need to know where the cloud conf is and how to consume it (ie teach it the INI schema, or yaml if we migrate to yaml?). This spreads a lot of ugly logic around the different components.

2) We store the zonal information on the Infrastructure object as well - any project that wishes to consume this information can pull the information as a Kube structured object as they would for any other resource, no special ini/yaml parsing logic required, as a structured go type straight away

While this enhancement doesn't cover those potential future use cases, I'm wondering how much extra work it is to store that information so that future projects that might need this information don't need to learn the format of the cloud config for vSphere and how to read INI files

Reply:
Do you think we need to add the zonal information at the start of this feature or can wait? I don't see any changes that would restrict us to adding that later, unless upgrades would become a problem?

Reply:
It's upgrades that concern me, if some clusters have the zone information and some don't, then future projects such as CAPI/MAPI zonal support have to implement a fallback to read from the cloud conf, which defeats the point of putting it in the infrastructure object. That or we have to tell users to configure the zone information in the infrastructure object